packages/esm-form-entry-app/src/app/form-data-source/form-data-source.service.ts

```diff
@@ -114,6 +117,56 @@ export class FormDataSourceService {
     );
   }
 
+  public getMostRecentObsValueBefore(date: string, patientUuid: string) {
+    type ObsResult = FHIRResource['resource'] & {
+      conceptUuid: string;
+      dataType?: string;
+    };
+
+    function isUuid(input: string) {
+      return input.length === 36;
+    }
+
+    const dataSource = {
+      obsValues: [],
+    };
+
+    const concepts = this.conceptResourceService.getAllConcepts();
+    concepts.subscribe((concepts) => {
+      const uuids = concepts.map((c) => c.uuid).join(',');
+
+      return this.observationResourceService.getMostRecentObsValues(date, uuids, patientUuid).subscribe((data) => {
+        const observations =
+          data?.entry?.map((entry) => {
+            const observation: ObsResult = {
+              ...entry.resource,
+              conceptUuid: entry.resource.code.coding.filter((c) => isUuid(c.code))[0]?.code,
+            };
+
+            if (entry.resource.hasOwnProperty('valueString')) {
+              observation.dataType = 'Text';
```

Comment:
I'm not exactly sure why this is necessary?

Reply:
I saw in some resources how they were formatting observations and I used the same approach but I will look closely and see if it is indeed necessary.

Reply:
Yeah, we use that in some formatting code, but since this is designed to be called in an expression, I don't think we have the same set of concerns... (In some cases we need to work out whether the Obs should be displayed as a textbox, a number input, etc).